# Model Training

## Table of Contents
- Evaluation Metrics Used
- Modelling
- Comparison of Models
- Results

<br>


## Evaluation Metrics Used

#### Accuracy
Accuracy provides a ratio of correctly predicted observations to the total observations. 
**Formula**:
$$\text{Accuracy} = \frac{\text{Number of Correct Predictions}}{\text{Total Number of Predictions}}$$

#### Confusion Matrix and Related Terms
The confusion matrix is a table layout that allows visualization of the performance of the algorithm, where each number in the matrix represents:
- **TP (True Positives)**: Correctly predicted positive observations.
- **TN (True Negatives)**: Correctly predicted negative observations.
- **FP (False Positives)**: Incorrectly predicted as positive.
- **FN (False Negatives)**: Incorrectly predicted as negative.

#### Precision, Recall, and F1-Score
- **Precision**:
$$ \text{Precision} = \frac{\text{TP}}{\text{TP} + \text{FP}} $$
Precision measures the accuracy of positive predictions.

- **Recall** (or Sensitivity or TPR):
$$ \text{Recall} = \frac{\text{TP}}{\text{TP} + \text{FN}} $$
Recall measures the ability of a model to find all the relevant cases (all positive samples).

- **F1-Score**:
$$ \text{F1-Score} = 2 \times \left( \frac{\text{Precision} \times \text{Recall}}{\text{Precision} + \text{Recall}} \right) $$
The F1-Score is the weighted average of Precision and Recall. Therefore, this score takes both false positives and false negatives into account. A high F1-score shows a model can classify the positive class correctly, while not misclassifying many negative classes as positive.

<br>

## Import data and train_test_split

In [23]:
import pandas as pd
df = pd.read_csv("data/cleaned_df_v2.csv", index_col=False)
df = df.iloc[: , 1:]

# convert date
df['year_month'] = pd.to_datetime(df['year_month'])
df['year'] = df['year_month'].dt.year
df['month'] = df['year_month'].dt.month

In [24]:
from sklearn.model_selection import train_test_split

features = ['year', 'month', 'us_state', 'make', 'phone_series',  'model_family','model_age_months', 
            'predecessor', 'successor', 'model', 'sku', 'sku_no_colour', 'colour', 'size', 'claims', 
            'ir_rate_swap', 'ir_rate_replacement', 'ir_rate_monthly', 'churn_rate'] # 19 features
target = 'closing_subs_monthly'

X = df[features]
y = df[target]

X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.3, random_state=3244
    )

X_train.head()

year  month    us_state     make           phone_series  \
6615   2023      6      Nevada    Apple              IPHONE 12   
9440   2023      5      Nevada    Apple              IPHONE XR   
1549   2022     12       Texas     Oppo               OPPO A90   
10717  2019      9      Nevada  Samsung  SAMSUNG GALAXY NOTE 9   
3034   2021      4  California  Samsung             GALAXY S21   

                  model_family  model_age_months              predecessor  \
6615   APPLE IPHONE 12 PRO MAX                31  APPLE IPHONE 11 PRO MAX   
9440           APPLE IPHONE XR                55           APPLE IPHONE X   
1549                  OPPO A94                21              First Model   
10717    SAMSUNG GALAXY NOTE 9                13              First Model   
3034   SAMSUNG GALAXY S21 PLUS                14  SAMSUNG GALAXY S20 PLUS   

                     successor                                model  \
6615   APPLE IPHONE 13 PRO MAX  IPHONE 12 PRO MAX 5G 256GB GRAPHITE   
9440           APPLE IPHONE 11                 IPHONE XR 64GB BLACK   
1549                Last Model                     A94 128GB SILVER   
10717               Last Model            GALAXY NOTE 9 128GB BLACK   
3034   SAMSUNG GALAXY S22 PLUS       GALAXY S21 PLUS 5G 256GB BLACK   

                    sku  sku_no_colour  colour  size  claims  ir_rate_swap  \
6615   IPH12PRM5G256GRY  IPH12PRM5G256    Gray   256     3.0           0.0   
9440         IPHXR64BLK        IPHXR64   Black    64     0.0           0.0   
1549          A94128SIL         A94128  Silver   128     0.0           0.0   
10717       NOTE9128BLK       NOTE9128   Black   128    16.0           0.0   
3034    GXYS21P5G256BLK   GXYS21P5G256   Black   256     0.0           0.0   

       ir_rate_replacement  ir_rate_monthly  churn_rate  
6615                   0.0         0.001585    0.049128  
9440                   0.0         0.000000    0.055838  
1549                   0.0         0.000000    0.073171  
10717                  0.0         0.002614    0.035778  
3034                   0.0         0.000000    0.027237

In [25]:
categorical_cols = ["us_state", "make", "phone_series", "model_family", "colour", "size"]

## RandomForestRegressor to predict closing_subs_monthly

In [26]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import numpy as np


# Define column transformer
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(drop='first'), categorical_cols)  
    ])

# Define the model pipeline
model_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('regressor', RandomForestRegressor(n_estimators=100, random_state=42)) 
])

# Model training
model_pipeline.fit(X_train, y_train)

# Predictions and evaluation
y_pred = model_pipeline.predict(X_test)

# Evaluate the model performance
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f'Root Mean Squared Error (RMSE): {rmse}')

Root Mean Squared Error (RMSE): 1870.7079620144557


RMSE of 1870.71 indicates that, on average, the model's predictions are off by about 1870.71 units.

